src/dune/docker.py

```diff
@@ -66,21 +70,58 @@ def get_container(self):
     def get_image(self):
         return self._image
 
-    def execute_docker_cmd(self, cmd):
-        with subprocess.Popen(['docker'] + cmd,
-                              stdout=subprocess.PIPE, stderr=subprocess.PIPE) as proc:
-            stdout, stderr = proc.communicate()
-            if self._cl_args.debug:
-                print('docker '+' '.join(cmd))
-                print(stdout.decode('UTF-8'))
-                print(stderr.decode('UTF-8'))
-        return [stdout.decode('UTF-8'), stderr.decode('UTF-8'), proc.poll()]
+    @staticmethod
+    def print_streams(stdout, stderr):
+        if stdout is None and stderr is None:
+            print('!! No stdout/stderr info captured...')
+            return
+
+        print('================ STDOUT ================')
+        print(stdout)
+        print('================ STDERR ================')
+        print(stderr)
+        print('========================================')
+
+    def execute_docker_cmd(self, cmd, *, check_status=True, capture_output=True):
+        """Execute the given docker command in the active container.
+
+        :param check_status: if command has a return code != 0 then raise an exception
+        :param capture_output: whether to capture stdout and stderr and return them or to
+                               print the streams normally
+        :return: (stdout, stderr, status_code) if captured, (None, None, status_code) otherwise
+        """
+        if capture_output:
+            with subprocess.Popen(['docker'] + cmd,
+                                  stdout=subprocess.PIPE, stderr=subprocess.PIPE) as proc:
+                stdout, stderr = proc.communicate()
+                stdout = stdout.decode('UTF-8')
+                stderr = stderr.decode('UTF-8')
+                status = proc.returncode
+
+        else:
+            with subprocess.Popen(['docker'] + cmd) as proc:
+                proc.communicate()
+                stdout = None
+                stderr = None
+                status = proc.returncode
+
+        if check_status and status != 0:
+            # some error happened, log it and fail
+            print(f'ERROR: docker {cmd}  -- returned: {status}')
+            self.print_streams(stdout, stderr)
+            raise docker_error
+
+        if self._cl_args.debug:
+            print('docker '+' '.join(cmd))
+            self.print_streams(stdout, stderr)
+
+        return (stdout, stderr, status)
 
     def file_exists(self, file_name):
-        return self.execute_cmd(['test', '-f', file_name])[2] == 0
+        return self.execute_cmd(['test', '-f', file_name], check_status=False)[2] == 0
```

Comment:
Why you have decided not to check status in file_exists, dir_exists and wallet_unlock?

Reply:
I did it because a non-zero status code means that the file/dir does not exist and is not a failure case, so we want to check it ourselves (this is what the end of the line: ` == 0` does)

Reply:
as for `wallet_unlock`, it is because unlocking an already unlocked wallet returns a non-zero code in cleos, however we don't want to fail in dune. I added some smarter error handling in the case we get a non-zero code and the error is not about the wallet being unlocked